beacon_node/http_api/src/block_id.rs

```diff
@@ -252,23 +255,39 @@ impl BlockId {
         }
     }
 
-    /// Return the `BlobSidecarList` identified by `self`.
-    pub fn blob_sidecar_list<T: BeaconChainTypes>(
+    #[allow(clippy::type_complexity)]
+    pub fn get_blinded_block_and_blob_list_filtered<T: BeaconChainTypes>(
         &self,
+        indices: BlobIndicesQuery,
         chain: &BeaconChain<T>,
-    ) -> Result<BlobSidecarList<T::EthSpec>, warp::Rejection> {
-        let root = self.root(chain)?.0;
-        chain
+    ) -> Result<
+        (
+            SignedBlindedBeaconBlock<T::EthSpec>,
+            BlobSidecarList<T::EthSpec>,
+            ExecutionOptimistic,
+            Finalized,
+        ),
+        warp::Rejection,
+    > {
+        let (root, execution_optimistic, finalized) = self.root(chain)?;
+        let block = match &self.0 {
+            CoreBlockId::Head => {
+                let (cached_head, _) = chain
+                    .canonical_head
+                    .head_and_execution_status()
+                    .map_err(warp_utils::reject::beacon_chain_error)?;
+                cached_head.snapshot.beacon_block.clone_as_blinded()
+            }
```

Comment:
Although it's nice to fetch the head block without having to go to the DB, there's a race condition here because the canonical head could change between the call to get the `root` and the call to load the `head_and_execution_status`. This race could result in occasionally returning the wrong data, like the blobs from slot `N` and the block from slot `N + 1`.

I think the simplest fix might be to just remove the special casing for the head here. Loading from the DB usually isn't too slow, because the DB does its own caching